src/controller/CHIPDeviceControllerSystemState.h

```diff
@@ -0,0 +1,146 @@
+/*
+ *
+ *    Copyright (c) 2021 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      DeviceControllerSystemState is a representation of all the runtime state
+ *      inside of CHIP that can be shared across Controllers and Commissioners.
+ *
+ *      The System State assumes it's being owned by and managed by the DeviceControllerFactory.
+ *      It will automatically shutdown the underlying CHIP Stack when its reference count
+ *      decreases to "1".
+ *
+ */
+
+#pragma once
+
+#include <controller/DeviceControllerInteractionModelDelegate.h>
+#include <protocols/secure_channel/MessageCounterManager.h>
+#include <transport/TransportMgr.h>
+#if CONFIG_DEVICE_LAYER
+#include <platform/CHIPDeviceLayer.h>
+#endif
+
+#if CONFIG_NETWORK_LAYER_BLE
+#include <ble/BleLayer.h>
+#endif
+
+namespace chip {
+
+namespace Controller {
+
+struct DeviceControllerSystemStateParams
+{
+    System::Layer * systemLayer = nullptr;
+    Inet::InetLayer * inetLayer = nullptr;
+#if CONFIG_NETWORK_LAYER_BLE
+    Ble::BleLayer * bleLayer = nullptr;
+#endif
+    DeviceTransportMgr * transportMgr                             = nullptr;
+    SessionManager * sessionMgr                                   = nullptr;
+    Messaging::ExchangeManager * exchangeMgr                      = nullptr;
+    secure_channel::MessageCounterManager * messageCounterManager = nullptr;
+    Transport::FabricTable * fabricTable                          = nullptr;
+    DeviceControllerInteractionModelDelegate * imDelegate         = nullptr;
+};
+
+// A representation of the internal state maintained by the DeviceControllerFactory
+// and refcounted by Device Controllers.
+// Expects that the creator of this object is the last one to release it.
+class DeviceControllerSystemState
```

Comment:
Is there any reason that the DeviceControllerSystemState can't just be a member of DeviceControllerFactory and have its lifetime bound to that?

Reply:
I wanted to allow for shutdown/startup of the system state when not in use.

Reply:
I guess I'm still confused.  Why then isn't this exposed as Init / Shutdown methods in the factory object?  That seems to be the pervasive pattern in the SDK.

Reply:
We still have the init/shutdown option on the Controller objects. During shutdown, the controller relinquishes interest in the system state and releases any devices it has setup and stops handling messages. 

Offering a shutdown/startup on the Factory object leads to complications with us not being able to enforce the shutdown without holding references to all controllers that were vended and individually shutting them down as well. 

I instead chose to automatically do this as controllers are shutdown individually. They drop their references to the system state and when the system is idle (i.e, no controller object is running) it just tears itself down. The factory then restarts it when it vends the next controller.

Reply:
OK, that makes sense.

Thank you.